EIPS/eip-overlay_tree.md

```diff
@@ -0,0 +1,108 @@
+  * The third phase begins when a sufficient number of consecutive blocks report the same value: the overlay tree is progressively merged back into the newly converted binary base trie. A constant number of entries are deleted from the overlay and inserted into the base trie.
+  * The fourth and final phase begins when the overlay trie is empty. The field holding its root is removed from the block header.
+
+## Motivation
+
+There is a long running interest in switching the state trie from a hexary format to a binary format, for reasons pertaining to proof and storage sizes. The conversion process poses a catch-up issue, caused by the sheer size of the full state: it can not be translated in a reasonable time (i.e. on the same order of magnitude as the block time). 
+
+## Specification
+The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).
+
+### Binary tries
+
+This EIP assumes that a binary trie is defined like the MPT, except that:
+
+  * The series of bytes in I₀ is seen as a series of _bits_ and so ∀i≤256, I₀[i] is the ith bit in key I₀
+  * The first item of an **extension** or a **leaf** is replacing nibbles with bits;
+  * A **branch** is a 2 item structure in which both items correspond to each of the two possible bit values for the keys at this point in their traversal;
+  * c(𝕴,i) ≡ RLP((u(0), u(1)) at a branch, where u(j) = n({I : I ∈ 𝕴 ⋀ I₀[i] = j}, i+1)
+
+### Phase 1
+
+Let _H₁_ be the previously agreed-upon block height at which phase 1 starts. For each block of height H₁ ≤ _h_ < H₂:
+
+  * Block headers contain a block header has a new Hₒ which is the _root of the overlay binary trie_
+  * Hᵣ ≡ P(H)ᵣ, i.e. the hexary trie root is the same as that of the block parent's. The hexary trie is referred to as the _hexary base trie_.
+
+The following is changed in the execution environment:
+
+  * Upon executing a _state read_, ϒ first searches for the address in the overlay trie. If the key can not be found there, ϒ then searches the hexary base trie as it did at block heights h' < H₁;
+  * Upon executing a _state write_, ϒ will insert or update the value into the overlay tree.
+
+A conversion process occurs in the background to turn the hexary trie into its binary equivalent. The end goal of this process is the calculation of the _root hash of the converted binary base trie_, denoted Hᵣ². The root of the hexary base trie is hereafter called Hᵣⁱ⁶.
+
+When a miner has calculated Hᵣ², it proceeds to phase 2.
+
+### Phase 2
+
+Phase 2 is the same as phase 1, except for the following:
+
+  * Hᵣ' ≡ Hᵣ²
+
+
+Phase 2 ends when a sufficient number of subsequent blocks have reported the same value for Hᵣ².
```

Comment:
Does this mean that before this happen, `hi2 = "donald_duck"` is a valid block? 
Shouldn't it say, instead

> .. when a sufficient number of subsequent blocks contain binary root hash. 

(assuming that only valid root hashes are 'valid')

Also, does this mean that it's acceptable for a miner to mine on top of a (non-verified) binary-hashed block, but omit the binary-root hash (a phase-one block)?

Could a valid block progression be: `p1, p1, p2, p2, p1, p2, p2,p1,p2` ?

Reply:
Any value for Hᵣ² is valid. What I'm trying to achieve here with the transition from phase 2 to phase 3 is some kind of vote: miners "vote" the value that they converted, through this Hᵣ² field. If there is an agreement on the same value of more than 51% of the network (the sequence of blocks is a proxy for that) then it is accepted as the "official" value.

If more than 51% of the network is advertising `donald_duck`, then a majority of the network has effectively killed it (let alone committed a copyright infringement, Disney's not kidding with these).

Can the client run by miner build on top of a non-verified Hᵣ²? Yes it can if they trust another party with their conversion. However, they would still need the state as soon as phase 3 starts, because it is needed to progressively merge the overlay back into the base trie. Since in phase 3 both Hᵣ² and Hₒ are part of the consensus, trusting a value without calculating the state is risky.

> Could a valid block progression be: p1, p1, p2, p2, p1, p2, p2,p1,p2 ?

Yes, in this case it means that clients don't agree on their conversion results. It will happen at the beginning when one miner is done converting. The votes should converge over time, unless one actor with most of the hashing power is malicious or made an error.

Reply:
So after discussing this on other media, there doesn't seem to be any advantages to the complexity of this approach. I have updated this document to present a simpler transition: it will unconditionally happen at block height h2, which is set far apart in the future to make sure that honest nodes are done with the conversion. Measurements indicate that this can be achieved in less than a day.